Mathlib/Algebra/FreeMonoid/Basic.lean

```diff
@@ -415,4 +423,41 @@ theorem length_reverse {a : FreeMonoid α} : a.reverse.length = a.length :=
 
 end Reverse
 
+section IsomorphicTypes
+
+variable {α β : Type*}
+
+/-- free monoids over isomorphic types are isomorphic -/
+@[to_additive "if two types are isomorphic, the additive free monoids over those types are
+isomorphic"]
+def congrEquiv (e : α ≃ β) : FreeMonoid α ≃* FreeMonoid β :=
+  MulEquiv.mk' ⟨FreeMonoid.map e.toFun, FreeMonoid.map e.invFun, fun _ => map_invFun_map_toFun_eq e,
+    fun _ => map_toFun_map_invFun_eq e⟩ (by simp [map_mul])
+
+@[to_additive (attr := simp)]
+theorem congrEquiv_one (e : α ≃ β) : congrEquiv e (1) = 1 := rfl
+
+@[to_additive (attr := simp)]
+theorem congrEquiv_of (e : α ≃ β) (a : α) : congrEquiv e (of a) = of (e.toFun a) := rfl
+
+@[to_additive (attr := simp)]
+theorem congrEquiv_mul (e : α ≃ β) (a b : FreeMonoid α) : congrEquiv e (a * b) = congrEquiv e a *
+    congrEquiv e b := by rw [map_mul]
+
+/-- given a function from β to α, convert a relation predicate on FreeMonoid α to
+have an underlying type of β -/
+@[to_additive "given a function from β to α, convert a relation predicate to
+have an underlying type of β"]
+def mapRel (f : β → α)(rel : FreeMonoid α → FreeMonoid α → Prop) : FreeMonoid β → FreeMonoid β  →
```

Comment:
For these next two definitions, could you please add `_iff` lemmas that state how they interact with generators?
```suggestion
def mapRel (f : β → α) (rel : FreeMonoid α → FreeMonoid α → Prop) : FreeMonoid β → FreeMonoid β  →
```

Reply:
I'm sorry, I can't quite figure out what you mean. I don't see how I could apply this to the FreeMonoid generators (it is entirely possible that none of the relations holds between elements of length one - for example, in the braid monoid), and there aren't generators for `f` or `rel`

If you could give me a hint, I'd be happy to add this :)

Reply:
this is deleted per Yury's suggestion :)